base_layer/core/src/proof_of_work/cuckaroo_pow.rs

```diff
@@ -0,0 +1,476 @@
+    let mut mini_buffer = 0u64;
+    let mut remaining = 64;
+    for el in uncompressed {
+        dbg!(el, bit_width, remaining, mini_buffer);
+
+        mini_buffer |= el << (64 - remaining);
+        dbg!(format!("mini_buffer: {:064b}", mini_buffer));
+        if bit_width < remaining {
+            remaining -= bit_width;
+        } else {
+            compressed[..8].copy_from_slice(&mini_buffer.to_le_bytes());
+            compressed = &mut compressed[8..];
+            mini_buffer = el >> remaining;
+            remaining = 64 + remaining - bit_width;
+        }
+    }
+    let mut remainder = compressed.len() % 8;
+    if remainder == 0 {
+        remainder = 8;
+    }
+    if mini_buffer > 0 {
+        compressed[..].copy_from_slice(&mini_buffer.to_le_bytes()[..remainder]);
+    }
+    dbg!(&target);
+    dbg!(format!("target: {:08b}", target[0]));
+    dbg!(format!("target: {:08b}", target[1]));
+    target
+}
+
+fn unpack_nonces(pow: &[u8], edge_bits: u8, expected_length: usize) -> Result<Vec<u64>, CuckarooVerificationError> {
+    let mut nonces = Vec::with_capacity(expected_length);
+    let node_mask = (1u64 << edge_bits) - 1;
+    let mut mini_buffer = 0u64;
+    let mut remaining = 64;
+    let mut bytes = pow.iter().copied();
+    while let Some(byte) = bytes.next() {
+        mini_buffer |= u64::from(byte) << (64 - remaining);
+        remaining -= 8;
+        while remaining <= 64 - edge_bits {
+            let nonce = mini_buffer & node_mask;
+            if nonce > node_mask {
+                return Err(CuckarooVerificationError::NonceTooLarge);
+            }
+            nonces.push(nonce);
+            mini_buffer >>= edge_bits;
+            remaining += edge_bits;
+        }
+    }
+
+    for n in nonces[expected_length..].iter() {
+        if *n != 0 {
+            return Err(CuckarooVerificationError::PowDataContainsNonZeroPadding);
+        }
+    }
+    Ok(nonces.into_iter().take(expected_length).collect())
+}
+
+fn verify(
+    siphash_keys: &[u64; 4],
+    nonces: &[u64],
+    cycle_length: usize,
+    edge_bits: u8,
+) -> Result<(), CuckarooVerificationError> {
+    let node_mask = (1u64 << edge_bits) - 1;
+    let mut uvs = Vec::with_capacity(cycle_length);
+    for i in 0..cycle_length {
+        if nonces[i] > node_mask {
+            return Err(CuckarooVerificationError::NonceTooLarge);
+        }
+        if i > 0 && nonces[i] <= nonces[i - 1] {
+            return Err(CuckarooVerificationError::NoncesNotAscending);
+        }
+
+        let edge = siphash_block(&siphash_keys, nonces[i], 21, true);
+        let u = edge & node_mask;
+        let v = (edge >> 32) & node_mask;
+
+        uvs.push((u, v));
+    }
+
+    // Verify the cycle from the edges
+    verify_from_edges(&uvs, cycle_length)
+}
+
+fn verify_from_edges(uvs: &[(u64, u64)], cycle_length: usize) -> Result<(), CuckarooVerificationError> {
+    let proof_size = uvs.len();
+    if proof_size != cycle_length {
```

Comment:
```suggestion
    if proof_size != cycle_length || cycle_length == 0 {
```
then we can safely `expect` later on